File: packages/go/protocol/transport_adapters_test.go
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 DeRec Alliance. All rights reserved.

package protocol

import (
	"errors"
	"reflect"
	"testing"
)

// recordingDialer records every endpoint it was asked to dial and refuses the
// ones named.
type recordingDialer struct {
	attempted []string
	refuse    map[string]bool
}

func newDialer(refuse ...string) *recordingDialer {
	set := make(map[string]bool, len(refuse))
	for _, uri := range refuse {
		set[uri] = true
	}
	return &recordingDialer{refuse: set}
}

func (d *recordingDialer) SendOne(endpoint Endpoint, _ []byte) error {
	d.attempted = append(d.attempted, endpoint.URI)
	if d.refuse[endpoint.URI] {
		return errors.New("refused " + endpoint.URI)
	}
	return nil
}

func endpoints(uris ...string) []Endpoint {
	out := make([]Endpoint, 0, len(uris))
	for _, uri := range uris {
		out = append(out, Endpoint{URI: uri, Protocol: 0})
	}
	return out
}

const a, b, c = "https://a.example", "https://b.example", "https://c.example"

// Delivering to every endpoint would send one authenticated message to the
// same peer several times, so the first success has to end the attempt.
func TestSequentialFailover_StopsAtTheFirstSuccess(t *testing.T) {
	d := newDialer()
	if err := (SequentialFailover{Dialer: d}).Send(endpoints(a, b), []byte("m")); err != nil {
		t.Fatalf("Send: %v", err)
	}
	if want := []string{a}; !reflect.DeepEqual(d.attempted, want) {
		t.Fatalf("attempted = %v, want %v", d.attempted, want)
	}
}

func TestSequentialFailover_AdvancesPastAFailingEndpoint(t *testing.T) {
	d := newDialer(a)
	if err := (SequentialFailover{Dialer: d}).Send(endpoints(a, b), []byte("m")); err != nil {
		t.Fatalf("Send: %v", err)
	}
	if want := []string{a, b}; !reflect.DeepEqual(d.attempted, want) {
		t.Fatalf("attempted = %v, want %v", d.attempted, want)
	}
}

// The order is the peer's and is not reinterpreted.
func TestSequentialFailover_UsesThePeersOrder(t *testing.T) {
	d := newDialer(a)
	if err := (SequentialFailover{Dialer: d}).Send(endpoints(a, b, c), []byte("m")); err != nil {
		t.Fatalf("Send: %v", err)
	}
	if want := []string{a, b}; !reflect.DeepEqual(d.attempted, want) {
		t.Fatalf("attempted = %v, want %v", d.attempted, want)
	}
}

func TestSequentialFailover_FailsOnlyWhenEveryEndpointDid(t *testing.T) {
	d := newDialer(a, b)
	if err := (SequentialFailover{Dialer: d}).Send(endpoints(a, b), []byte("m")); err == nil {
		t.Fatal("expected an error when no endpoint accepted the message")
	}
	if want := []string{a, b}; !reflect.DeepEqual(d.attempted, want) {
		t.Fatalf("attempted = %v, want %v", d.attempted, want)
	}
}

func TestSingleEndpointTransport_IgnoresTheRest(t *testing.T) {
	d := newDialer()
	if err := (SingleEndpointTransport{Dialer: d}).Send(endpoints(a, b), []byte("m")); err != nil {
		t.Fatalf("Send: %v", err)
	}
	if want := []string{a}; !reflect.DeepEqual(d.attempted, want) {
		t.Fatalf("attempted = %v, want %v", d.attempted, want)
	}
}

// The distinguishing property: where SequentialFailover would recover, this
// reports the failure. That is the cost of choosing it, made observable.
func TestSingleEndpointTransport_DoesNotFallBack(t *testing.T) {
	d := newDialer(a)
	if err := (SingleEndpointTransport{Dialer: d}).Send(endpoints(a, b), []byte("m")); err == nil {
		t.Fatal("expected an error: the only endpoint tried was refused")
	}
	if want := []string{a}; !reflect.DeepEqual(d.attempted, want) {
		t.Fatalf("attempted = %v, want %v", d.attempted, want)
	}
}

// Both adapters satisfy the interface the builder takes.
func TestAdaptersAreTransports(t *testing.T) {
	var _ Transport = SequentialFailover{Dialer: newDialer()}
	var _ Transport = SingleEndpointTransport{Dialer: newDialer()}
}

File: packages/react-native/__tests__/transport-adapters.test.ts
import { sequentialFailover, singleEndpointTransport } from '../src/types';
import type { SendOne } from '../src/types';

const a = { uri: 'https://a.example', protocol: 'https' };
const b = { uri: 'https://b.example', protocol: 'https' };
const c = { uri: 'https://c.example', protocol: 'https' };

const message = new Uint8Array([1, 2, 3]);

/** Records every endpoint it was asked to dial, and refuses the ones named. */
function recorder(...refuse: string[]): { dialer: SendOne; attempted: string[] } {
  const attempted: string[] = [];
  const dialer: SendOne = async (endpoint) => {
    attempted.push(endpoint.uri);
    if (refuse.includes(endpoint.uri)) {
      throw new Error(`refused ${endpoint.uri}`);
    }
  };
  return { dialer, attempted };
}

describe('sequentialFailover', () => {
  it('stops at the first success rather than delivering twice', async () => {
    const { dialer, attempted } = recorder();
    await sequentialFailover(dialer).send([a, b], message);
    expect(attempted).toEqual([a.uri]);
  });

  it('advances past a failing endpoint', async () => {
    const { dialer, attempted } = recorder(a.uri);
    await sequentialFailover(dialer).send([a, b], message);
    expect(attempted).toEqual([a.uri, b.uri]);
  });

  it('tries the order the peer offered, not one of its own', async () => {
    const { dialer, attempted } = recorder(a.uri);
    await sequentialFailover(dialer).send([a, b, c], message);
    expect(attempted).toEqual([a.uri, b.uri]);
  });

  it('rejects only when every endpoint failed', async () => {
    const { dialer, attempted } = recorder(a.uri, b.uri);
    await expect(sequentialFailover(dialer).send([a, b], message)).rejects.toThrow();
    expect(attempted).toEqual([a.uri, b.uri]);
  });
});

describe('singleEndpointTransport', () => {
  it('uses the first endpoint and ignores the rest', async () => {
    const { dialer, attempted } = recorder();
    await singleEndpointTransport(dialer).send([a, b], message);
    expect(attempted).toEqual([a.uri]);
  });

  it('does not fall back where sequentialFailover would recover', async () => {
    const { dialer, attempted } = recorder(a.uri);
    await expect(singleEndpointTransport(dialer).send([a, b], message)).rejects.toThrow();
    expect(attempted).toEqual([a.uri]);
  });
});
